File: AAD/p6.py
import random

# Deterministic partition function using the last element as pivot
def deterministic_partition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]  # Swap elements

    arr[i + 1], arr[high] = arr[high], arr[i + 1]  # Swap pivot to its correct position
    return i + 1

# Randomized partition function that chooses a random pivot
def randomized_partition(arr, low, high):
    random_index = random.randint(low, high)
    arr[random_index], arr[high] = arr[high], arr[random_index]  # Swap random pivot with the last element
    return deterministic_partition(arr, low, high)

# Deterministic Quick Sort
def deterministic_quick_sort(arr, low, high):
    if low < high:
        pi = deterministic_partition(arr, low, high)
        deterministic_quick_sort(arr, low, pi - 1)
        deterministic_quick_sort(arr, pi + 1, high)

# Randomized Quick Sort
def randomized_quick_sort(arr, low, high):
    if low < high:
        pi = randomized_partition(arr, low, high)
        randomized_quick_sort(arr, low, pi - 1)
        randomized_quick_sort(arr, pi + 1, high)

# Main function to test the sorting algorithms
if __name__ == "__main__":
    n = int(input("Enter the number of elements in the array: "))
    arr = list(map(int, input("Enter the elements of the array:\n").split()))

    # Deterministic Quick Sort
    arr1 = arr[:]
    deterministic_quick_sort(arr1, 0, n - 1)
    print("Deterministic Quick Sort Result:")
    print(" ".join(map(str, arr1)))

    # Randomized Quick Sort
    arr2 = arr[:]
    randomized_quick_sort(arr2, 0, n - 1)
    print("Randomized Quick Sort Result:")
    print(" ".join(map(str, arr2)))

File: AAD/p6.cpp
#include <iostream>
#include <vector>
#include <cstdlib> // For rand()
#include <ctime>   // For seeding rand() with time

using namespace std;

// Deterministic partition function using the last element as pivot
int deterministicPartition(vector<int> &arr, int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++)
    {
        if (arr[j] <= pivot)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Randomized partition function that chooses a random pivot
int randomizedPartition(vector<int> &arr, int low, int high)
{
    int randomIndex = low + rand() % (high - low + 1);
    swap(arr[randomIndex], arr[high]);
    return deterministicPartition(arr, low, high);
}

// Deterministic Quick Sort
void deterministicQuickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int pi = deterministicPartition(arr, low, high);
        deterministicQuickSort(arr, low, pi - 1);
        deterministicQuickSort(arr, pi + 1, high);
    }
}

// Randomized Quick Sort
void randomizedQuickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int pi = randomizedPartition(arr, low, high);
        randomizedQuickSort(arr, low, pi - 1);
        randomizedQuickSort(arr, pi + 1, high);
    }
}

int main()
{
    srand(time(0)); // Seed for random number generation

    int n;
    cout << "Enter the number of elements in the array: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Enter the elements of the array:\n";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }

    // Deterministic Quick Sort
    vector<int> arr1 = arr;
    deterministicQuickSort(arr1, 0, n - 1);
    cout << "Deterministic Quick Sort Result:\n";
    for (int x : arr1)
        cout << x << " ";
    cout << endl;

    // Randomized Quick Sort
    vector<int> arr2 = arr;
    randomizedQuickSort(arr2, 0, n - 1);
    cout << "Randomized Quick Sort Result:\n";
    for (int x : arr2)
        cout << x << " ";
    cout << endl;

    return 0;
}
